### PR TITLE
Rewrite websocket

### PR DESCRIPTION
**Pull request by PatriceJiang**
Rewrite WebSocket with `message channel`. 
- lock
The original implementation use `mutex` a lot. most fields in `WebSocket` class are guarded by mutexes.
The problem is that it's hard to understand and difficult to maintain. like deadlock issues are hard to track. 
New implementation use `message channel` to communicate between threads, few mutexes is used, and it is even safer. 
- memory
Data class `WsMessage`  and `WebSocketFrame` are poorly managed and they are merged into`NetDataPack`.
smart pointers are used to improve reliability. 
- thread safety
`WebSocket` APIs are safe not limited to the main thread. 
- latency
`thread::sleep(3ms)` between `lws_service` is omitted by using libuv

### CODE REVIEW

File: cocos/base/loop/Looper.h [hunk 1]
```diff
@@ -0,0 +1,493 @@
...
+                }
+                else {
+                    auto &&front = _pendingFns.popFront();
+                    handleFn(front._data);
+                }
+            }
+
+            while (_pendingEvents.size() > 0)
+            {
+                SeqItem<LoopEvent> item = _pendingEvents.popFront();
```
Comment (crazyhappygame) [inline cocos/base/loop/Looper.h:449]: Why there is no lock here and there is lock in while loop above?

File: cocos/base/loop/Collections.h [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+            void popBack() { _CC_LOOP_DEFINE_LOCK_GUARD_; _data.pop_back(); }
+            T popFront() { _CC_LOOP_DEFINE_LOCK_GUARD_; assert(_data.size() > 0); T x = _data.front(); _data.pop_front(); return x; }
+
+            const T &front() { _CC_LOOP_DEFINE_LOCK_GUARD_; return _data.front(); }
+            const T &back() { _CC_LOOP_DEFINE_LOCK_GUARD_; return _data.back(); }
+
+            void clear() { _CC_LOOP_DEFINE_LOCK_GUARD_; _data.clear(); }
+
+            size_t size() const { return _data.size(); }
```
Comment (crazyhappygame) [inline cocos/base/loop/Collections.h]: why size without lock?

File: cocos/base/loop/Looper.h [hunk 3]
```diff
@@ -0,0 +1,493 @@
...
+        void Looper<LoopEvent>::wait(const std::function<void()> &fn, int timeoutMS)
+        {
+            if (isCurrentThread())
+            {
+                _pendingFns.pushBack(SeqItem<DispatchF>(genSeq(), fn));
+                onNotify();
+            }
+            else
+            {
+                std::condition_variable cv;
```
Comment (crazyhappygame) [inline cocos/base/loop/Looper.h]: Could you explain this part o code?
What is a point tocreate local mutex conditional variable lock and than pass reference and than wait?
It seems like when wait_for will expire we may have rice condition

**Review by crazyhappygame — CHANGES_REQUESTED** [review]

File: cocos/base/loop/Looper.h (hunk 1 above)
Comment (PatriceJiang) [reply cocos/base/loop/Looper.h:449]: `popFront` will be invoked only in `onNotify()`, one threaded. `item` will not be modified.

File: cocos/base/loop/Collections.h (hunk 2 above)
Comment (PatriceJiang) [reply cocos/base/loop/Collections.h]: `size()` itself is threadsafe, lock make no point here.

File: cocos/base/loop/Looper.h (hunk 3 above)
Comment (PatriceJiang) [reply cocos/base/loop/Looper.h]: `lock2` is not used, I will drop it.

Comment (PatriceJiang) [reply cocos/base/loop/Looper.h]: >It seems like when wait_for will expire we may have rice condition.

 yes. main thread should not be frozen forever even if it happens, and timeout value should be chosen properly.

File: cocos/base/loop/Collections.h (hunk 2 above)
Comment (crazyhappygame) [reply cocos/base/loop/Collections.h]: No, size of std::list (and any other and container) is not thread safe.
If in thread1 size is changed and thread2 check size this is race condition.

**Review by crazyhappygame — CHANGES_REQUESTED** [review]

File: cocos/base/loop/Looper.h (hunk 3 above)
Comment (crazyhappygame) [reply cocos/base/loop/Looper.h]: I am so sorry I do not understand it....
Why we need mutex mtx as local function variable?

File: cocos/base/loop/Looper.h (hunk 1 above)
Comment (crazyhappygame) [reply cocos/base/loop/Looper.h:449]: problem is with size.
Size is not thread safe

**Review by crazyhappygame — CHANGES_REQUESTED** [review]

File: cocos/base/loop/Looper.h (hunk 1 above)
Comment (PatriceJiang) [reply cocos/base/loop/Looper.h:449]: `size()` returns `size_t` value, it is 32-bits wide on 32-bits platform and 64-bits wide on 64-bits platform. 
single read (or write) is an atomic operation.

File: cocos/base/loop/Collections.h (hunk 2 above)
Comment (PatriceJiang) [reply cocos/base/loop/Collections.h]: add lock inside `size()` dose not help.

File: cocos/base/loop/Looper.h (hunk 3 above)
Comment (PatriceJiang) [reply cocos/base/loop/Looper.h]: the mutex is used in `cv.wait` and it will only be invoked once locally.

File: cocos/base/loop/Looper.h (hunk 1 above)
Comment (crazyhappygame) [reply cocos/base/loop/Looper.h:449]: That was (maybe) true in past for single core processors. Currently with mulitocore processors with multiple level or caches it is not true any more. E.g. without mutex second thread(on second core) may not see that size was changed

File: cocos/base/loop/LoopRunable.h [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+            void stop();
+            time_point<high_resolution_clock> expectTime()
+            {
+                return _startTime + milliseconds(_intervalMS * _updateTimes);
+            }
+        private:
+            Loop * _task = nullptr;
+            uv_loop_t *_uvLoop = nullptr;
+            uv_timer_t _uvTimer;
+            int64_t _intervalMS = 3000LL;
```
Comment (crazyhappygame) [inline cocos/base/loop/LoopRunable.h:58]: Pleas use std::chrono here and in any other places when time/duration is used

File: cocos/base/loop/Looper.h [hunk 5]
```diff
@@ -0,0 +1,501 @@
...
+
+            uv_loop_t *getUVLoop() { assert(_initialized); return _uvLoop; };
+
+        private:
+            void notify();
+            void onNotify();
+            void onStop();
+            void onRun();
+
+#if defined(__GNUC__)
```
Comment (crazyhappygame) [inline cocos/base/loop/Looper.h:122]: Why this is needed

File: cocos/base/loop/Looper.h (hunk 3 above)
Comment (crazyhappygame) [reply cocos/base/loop/Looper.h]: 1. But why for timeoutMS==0 add something to _pendingFns in thread1 and than wait inside thread1 until thread2 will finish?
2. For timeoutMS>0 is when cv.wait_for(lock, milliseconds(timeoutMS)); will expire than std::condition_variable cv; but cv is passed as reference to lambda [&cv  this is crash/undefined behaviur....

File: cocos/base/loop/Looper.h (hunk 1 above)
Comment (crazyhappygame) [reply cocos/base/loop/Looper.h:449]: 1. OK. Size is fixed. Thank you! By the way e.g. here is https://baptiste-wicht.com/posts/2012/03/cp11-concurrency-tutorial-part-2-protect-shared-data.html there is simple example which you can run multiple times and see that int is really not atomic 
2. But I see another problem 
```
while(_pendingEvents.size() > 0)
//Between those two lines another thread may add remove item from  _pendingEvents and we have undefined behavior
SeqItem<LoopEvent> item = _pendingEvents.popFront();
```

File: cocos/base/loop/Looper.h [hunk 6]
```diff
@@ -0,0 +1,501 @@
...
+            ThreadSafeMapArray<std::string, EventCF> _callbackMap;
+            ThreadSafeQueue<SeqItem<LoopEvent>> _pendingEvents;
+            ThreadSafeQueue<SeqItem<DispatchF>> _pendingFns;
+            
+#if defined(__GNUC__)
+            volatile uint64_t _eventFnSeq = 0;
+#else
+            std::atomic_uint64_t _eventFnSeq{ 0 };
+#endif 
+            bool _forceStoped = false;
```
Comment (crazyhappygame) [inline cocos/base/loop/Looper.h:145]: I think all those members should be changed to atomic protected by mutex

**Review by crazyhappygame — CHANGES_REQUESTED** [review]
@minggo @PatriceJiang I did not spend much time on review but current implementation have many threads problems. Such problems are hard to reproduce fix and debug ....

I think that even if we fix all those problems we will have code which is much more complicated than current implementation with more or less the same number of locks and atomics and performance problems...

Please double check current implementation. All variables (even bools and integers) must be protected by mutex (or change to atomic) if there is at least one write/update in multi-threaded code.

I checked current WebSocket.cpp and I think is much simpler to understand what's going on than this PR. 

Instead of rewriting everything I propose to fix/simplify current WebSocket.cpp implementation. 
uv is very nice library but I doubt if it fix all our current poblems.

**Comment by minggo** [comment]
@crazyhappygame ok, i will take more time on this PR. The biggest purpose of this PR is simplify the codes and make it more safe. If you think it is more complex, then it may have problem.

File: cocos/base/loop/Looper.h (hunk 3 above)
Comment (PatriceJiang) [reply cocos/base/loop/Looper.h]: my fault, I will use shared_ptr to protect `cv`

**Comment by PatriceJiang** [comment]
It is too complicated and risky, so closed.

**Pull request closed without merging**